### PR TITLE
feat: Generate Communication Invitation

### LINKED ISSUE #13: Generate Communication Invitation
DIDComm generates the connection invitation and it's upto Agent to share it (it could be through HTTP, QR Code, Bluetooth).

### PR DESCRIPTION
**Pull request by rolsonquadras**
DIDComm generates the connection invitation and it's upto Agent to share it (it could be through HTTP, QR Code, Bluetooth).

Closes #13 

Signed-off-by: Rolson Quadras <rolson.quadras@securekey.com>

### CODE REVIEW

**Comment by llorllale** [comment]
@rolsonquadras let's first discuss the design we're going for in #13

File: pkg/connection/connection.go [hunk 1]
```diff
@@ -31,46 +31,46 @@ func NewDIDComm(transport transport.OutboundTransport) *DIDComm {
 	return &DIDComm{transport: transport}
 }
 
-// SendInviteWithPublicDID sends the invite with public DID
-func (comm *DIDComm) SendInviteWithPublicDID(id, label, did string) error {
-	if id == "" && did == "" {
-		return errors.New("id and DID are necessary to send the Invite message")
+// GenerateInviteWithPublicDID generates the invite with public DID
+func GenerateInviteWithPublicDID(id, label, did string) (string, error) {
```
Comment (llorllale) [inline pkg/connection/connection.go]: @rolsonquadras I feel like this function's name and description should reflect more accurately what particular message this is. This is specifically the "invitation to exchange" message. So we can rename this function to more accurately reflect that (`NewExchangeInvite`?).

File: pkg/connection/connection.go [hunk 2]
```diff
@@ -31,46 +31,46 @@ func NewDIDComm(transport transport.OutboundTransport) *DIDComm {
...
-	return comm.transport.Send(base64.URLEncoding.EncodeToString(invitationJSON), "https://localhost:8090") // TODO need to add destination url here
+	return base64.URLEncoding.EncodeToString(invitationJSON), nil
 }
 
-// SendInviteWithKeyAndURLEndpoint sends the invite with recipient key and URL endpoint
-func (comm *DIDComm) SendInviteWithKeyAndURLEndpoint(id, label string, recipientKeys []string, serviceEndpoint string, routingKeys []string) error {
-	if id == "" {
-		return errors.New("id is necessary to send the Invite message")
+// GenerateInviteWithKeyAndURLEndpoint generates the invite with recipient key and URL endpoint
+func GenerateInviteWithKeyAndURLEndpoint(id, label string, recipientKeys []string, serviceEndpoint string, routingKeys []string) (string, error) {
```
Comment (llorllale) [inline pkg/connection/connection.go]: @rolsonquadras perhaps for another task but: we should have the first function above take in these optional parameters (optionally). Same for the function below. You can do it now, or leave a `TODO`

**Comment by rolsonquadras** [comment]
> @rolsonquadras let's first discuss the design we're going for in #13

@llorllale I have replied to your comments on the issue and also had offline discussion. Let me know if you have any questions.

File: pkg/connection/connection.go (hunk 1 above)
Comment (rolsonquadras) [reply pkg/connection/connection.go]: Updated the comments. But I feel current function name reflects better for a lib. It should be for the agent to say it's New/Existing or something else. The purpose of this function is to generate/return the invitation string for args passed.

File: pkg/connection/connection.go (hunk 2 above)
Comment (rolsonquadras) [reply pkg/connection/connection.go]: @llorllale Sorry, I don't understand this comment. Could you please elaborate more or add some snippet?

**Comment by rolsonquadras** [comment]
> @rolsonquadras left a couple of comments

@llorllale Replied to your comments.

File: pkg/connection/connection.go (hunk 2 above)
Comment (Baha-sk) [reply pkg/connection/connection.go]: the function has too many args.. create a struct  instead or variadic opts (id, label string, opts ...Option)

Comment (llorllale) [reply pkg/connection/connection.go]: @rolson looking closer, it seems the two functions below are doing the exact same thing. I wonder why they shouldn't be collapsed into one instead.

Comment (rolsonquadras) [reply pkg/connection/connection.go]: @Baha-sk agreed. Added a struct.

Comment (rolsonquadras) [reply pkg/connection/connection.go]: @llorllale The initial idea was to have 2 clear function names indicating one for DID service end point and another for URL endpoints for better segregation. But looking at it again, it makes sense to have one function. I have merged this into one now.

File: pkg/connection/connection.go (hunk 1 above)
Comment (llorllale) [reply pkg/connection/connection.go]: @rolsonquadras the name of the function is still misleading. "Public DID" is a loaded term - should not be in the name. I strongly suggest something like `NewDidExchangeInvite` name or something similar

Comment (rolsonquadras) [reply pkg/connection/connection.go]: @llorllale One of the invitation forms include a public DID https://github.com/hyperledger/aries-rfcs/tree/master/features/0023-did-exchange#0-invitation-to-exchange. To reflect this, the function is named such.

**Comment by rolson** [comment]
you guys keep pinging the wrong "rolson"

**Pull request merged**